### PR TITLE
Add type inference for Select-Object command

### LINKED ISSUE #7170: Add type inference for Select-Object
Steps to reproduce
------------------

```powershell
get-childitem | Select-Object -first 1 | foreach <Tab>
```

Expected behavior
-----------------

```none
Tab completion of the properties of the pipeline type
```

Actual behavior
---------------

```none
No completion occurs
```

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.0.2
PSEdition                      Core
GitCommitId                    v6.0.2
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0                                                                                                                                                                    
```

### PR DESCRIPTION
**Pull request by powercode**
## PR Summary

Add type inference for Select-Object command

Fixes #7170

## PR Checklist

- [x] [PR has a meaningful title]
- [x] [Summarized changes] 
This change add type inference for Select-Object in the cases when no projection is done.
I.E. when neither `-Property` or `-ExcludeProperty` is specified. 
- [x] [Change is not breaking]
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header] 
- [x] This PR is ready to merge and is not [Work in Progress]
- **User-facing changes**
    - [x] Not Applicable
    - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed]
    - [ ] [Add `[feature]` if the change is significant or affects feature tests]

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 1]
```diff
@@ -157,9 +157,9 @@ public bool TryGetRepresentativeTypeNameFromExpressionSafeEval(ExpressionAst exp
             {
                 return false;
             }
-            object value;
+
             return expression != null &&
-                   SafeExprEvaluator.TrySafeEval(expression, ExecutionContext, out value) &&
+                   SafeExprEvaluator.TrySafeEval(expression, ExecutionContext, out var value) &&
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: The PR contains so many style changes that makes it unreasonably large.
Let's leave only significant changes - it will greatly facilitate and speed up the code viewing. Please move style changes in another PR.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 2]
```diff
@@ -930,68 +918,91 @@ private IEnumerable<PSTypeName> InferTypesFromForeachCommand(PseudoBindingInfo p
...
+                }
+            }
+        }
+
+        private void InferTypesFromSelectCommand(PseudoBindingInfo pseudoBinding, CommandAst commandAst, List<PSTypeName> inferredTypes)
+        {
+            if (pseudoBinding.BoundArguments.TryGetValue("Property", out _)
+               || pseudoBinding.BoundArguments.TryGetValue("ExcludeProperty", out _))
+            {
+                return;
```
Comment (BrucePay) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: If the argument to `-Property` is a set of literal strings (most common case), then you could infer the "type" of the output object from them.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 3]
```diff
@@ -777,18 +780,17 @@ object ICustomAstVisitor.VisitFileRedirection(FileRedirectionAst fileRedirection
             return TypeInferenceContext.EmptyPSTypeNameArray;
         }
 
-        private IEnumerable<PSTypeName> InferTypesFrom(CommandAst commandAst)
+        private void InferTypesFrom(CommandAst commandAst, List<PSTypeName> inferredTypes)
         {
```
Comment (BrucePay) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: Why the change from IEnumerable to List?

**Review by BrucePay — CHANGES_REQUESTED** [review]
Over all it looks good. I'm still a bit curious about the change from `IEnumerable` -> `List`.  Also, i think the comments that talk about 'yield' need to be updated.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: I don't understand too why we add the allocations.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs (hunk 2 above)
Comment (powercode) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: Yes. I discussed this a bit with Jason. 
We don't currently have a good representation for this, but I would like to have that.
Same thing when you create a PSCustomObject from a hashtable.

But maybe something like "PSObject#Prop1/Prop2/Prop3"

It wouldn't say anything about the types of the properties, but that could maybe be encoded as well.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs (hunk 3 above)
Comment (powercode) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: Mostly debugability. It was a nightmare with all the nested state machines.

But since you have an allocation and virtual dispatch on every generated statemachine class that is generated when we use yield, my guess is that we actually have fewer allocations now.

And in general, I think the code got simpler. Several instances of loops that just yielded the loop object that just disappeared.

If you want to do the exercise, try debugging a tab completion before and after the change.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 4]
```diff
@@ -414,8 +410,7 @@ private static bool IsMemberHidden(object member)
         private static bool IsConstructor(object member)
         {
             var psMethod = member as PSMethod;
-            var methodCacheEntry = psMethod?.adapterData as DotNetAdapter.MethodCacheEntry;
-            return methodCacheEntry != null && methodCacheEntry.methodInformationStructures[0].method.IsConstructor;
+            return psMethod?.adapterData is DotNetAdapter.MethodCacheEntry methodCacheEntry && methodCacheEntry.methodInformationStructures[0].method.IsConstructor;
```
[the remark this answers is not in the document]
Comment (powercode) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: Agree! A bit to fast there :)

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: Yes, debugging can be a headache here.

As for allocations my understanding is that it depends on how we access the object - if we use IEnumerable interface we allocate in both cases, if we cast to a collection we don't allocate I believe.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 5]
```diff
@@ -350,8 +347,7 @@ private static bool TryGetRepresentativeTypeNameFromValue(object value, out PSTy
             type = null;
             if (value != null)
             {
-                var list = value as IList;
-                if (list != null && list.Count > 0)
+                if (value is IList list && list.Count > 0)
                 {
```
[the remark this answers is not in the document]
Comment (powercode) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: fixing.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs (hunk 3 above)
Comment (powercode) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs]: But there is no underlying collection to cast to. All the methods here that yield only has the generated state machine classes.

**Comment by powercode** [comment]
I don't think the Travis build failure has anything to do with this PR.

**Review by BrucePay — APPROVED** [review]

**Comment by iSazonov** [comment]
@powercode As for CodeFactor issues.
Many thanks that you fix ALL issues!!!

Really our policy is flexible:
- we don't ask fix all issues - only ones in new and changed code
- if you is _very busy_ you can ask maintainers to skip CodeFactor issues - we'll fix them ourselves later.
- if you do many changes it is better fix style issues in separate PR or add first separate commit with style changes.

Thanks again for your contribution!

**Comment by iSazonov** [comment]
Restart CI Travis Linux.

**Comment by anmenaga** [comment]
@TravisEz13 this looks ready; pls take a look when you can.
